pandas/core/algorithms.py

```diff
@@ -630,6 +634,9 @@ def factorize(
         uniques, codes = safe_sort(
             uniques, codes, na_sentinel=na_sentinel, assume_unique=True, verify=False
         )
+    if dropna is False and (codes == na_sentinel).any():
+        uniques = np.append(uniques, [np.nan])
```

Comment:
hmm, this should be a dtype appropriate for the dtype

Reply:
i think `dtype` is defined below in `_construct_data`? also, if added, for categorical values, will get `ValueError: Categorial categories cannot be null` error.

Reply:
right so instead of adding a null to the categories like you are doing, you just add the appropriate -1 entries in the codes which automatically handle the nullness

Reply:
emm, this has to be a `None` or `NA` value in there. otherwise, output of uniques does not have na value, and also, my python will crash with `FatalError` somehow